dcmotor.1
.\" -*- nroff -*-
.\" stepper is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License version 2
.\" as published by the Free Software Foundation.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; see the file COPYING.  If not, write to
.\" the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH DCMOTOR 1 "Dec 17, 2019"
.\" Please update the above date whenever this man page is modified.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins (default)
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
dcmotor \- libbbmotor DC motor example
.SH SYNOPSIS
.B dcmotor
.RI [ option... ]
.SH DESCRIPTION
\fBdcmotor\fP performs a CW speed ramp up on DC motor channel 1 on a
Seeed BeagleBone Motor Bridge Cape. The speed is changed on every second
by half of a procent.
The main goal of \fBdcmotor\fP is a living documentation of the dcmotor
functionality of \fBlibbbmotor\fP.
.PP
.SH OPTIONS
.TP
\fB\-d\fR, \fB\-\-duty\fR=\fIval\fR
Set duty to <val> milliseconds
.TP
\fB\-p\fR, \fB\-\-priority\fR=\fIprio\fR
Set SCHED_FIFO priority <prio>
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Print more information
.SH AUTHOR
Benedikt Spranger


ChangeLog
2019-12-17 23:21:05 +0100 Benedikt Spranger 

	* Add missing package to .travis (HEAD -> master, origin/master, github/master)

2019-12-12 17:26:28 +0100 Benedikt Spranger 

	* Add dcmotor example application

2019-12-12 19:08:47 +0100 Benedikt Spranger 

	* Add DC motor functions

2019-12-12 17:51:33 +0100 Benedikt Spranger 

	* Set default PWM frequency

2019-12-12 17:11:44 +0100 Benedikt Spranger 

	* Add stepper motor release call to demo application

2019-12-12 16:26:19 +0100 Benedikt Spranger 

	* Introduce motor port locking

2019-12-12 12:18:06 +0100 Benedikt Spranger 

	* Add pthread library test

2019-12-12 15:44:42 +0100 Benedikt Spranger 

	* Add stepper release function

2019-12-12 15:17:23 +0100 Benedikt Spranger 

	* Fix return values

2019-12-12 15:06:42 +0100 Benedikt Spranger 

	* Introduce GUARD_TIME constant

2019-11-18 17:35:32 +0100 Benedikt Spranger 

	* Release v0.1.2 (tag: v0.1.2)

2019-11-18 17:30:57 +0100 Benedikt Spranger 

	* Remove autogenerated .pot file on maintainer-clean

2019-11-18 15:50:48 +0100 Benedikt Spranger 

	* stepper: Adopt man page

2019-11-18 15:11:56 +0100 Benedikt Spranger 

	* stepper: Parameterize

2019-11-18 12:51:13 +0100 Benedikt Spranger 

	* stepper: Rework application output

2019-11-18 12:35:15 +0100 Benedikt Spranger 

	* Fix config.h include

2019-07-20 19:03:54 +0200 Benedikt Spranger 

	* Create docs

2019-07-20 17:16:34 +0200 Benedikt Spranger 

	* Release v0.1.1 (tag: v0.1.1)

2019-07-20 17:35:41 +0200 Benedikt Spranger 

	* Add manpages for the example applications

2019-07-20 15:15:42 +0200 Benedikt Spranger 

	* Encapsulate internal functions

2019-07-20 14:47:47 +0200 Benedikt Spranger 

	* Add servo functions

2019-07-20 14:32:31 +0200 Benedikt Spranger 

	* Split functionality

2019-07-20 13:56:10 +0200 Benedikt Spranger 

	* Add some settle time to reset

2019-07-19 03:52:38 +0200 Benedikt Spranger 

	* Add travis build status

2019-07-19 03:04:28 +0200 Benedikt Spranger 

	* Add travis configuration

2019-07-19 00:55:23 +0200 Benedikt Spranger 

	* Use SPDX license identifier (tag: v0.1.0)

2019-07-18 22:18:25 +0200 Benedikt Spranger 

	* Enhance documentation

2019-07-18 22:06:37 +0200 Benedikt Spranger 

	* Add GPIO handling

2019-07-18 18:08:17 +0200 Benedikt Spranger 

	* Add initial documentation

2019-07-18 15:03:47 +0200 Benedikt Spranger 

	* Initial commit



dcmotor.c
// SPDX-License-Identifier: GPL-2.0
// Copyright (C) 2019 Benedikt Spranger <b.spranger@linutronix.de>

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <argp.h>
#include <errno.h>
#include <sched.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include <sys/mman.h>

#include "bbmotor.h"

#if ENABLE_NLS
#include <libintl.h>
#define _(Text) gettext (Text)
#else
#define textdomain(Domain)
#define _(Text) Text
#endif
#define N_(Text) Text

static int stop;

static error_t parse_opt (int key, char *arg, struct argp_state *state);
static void show_version (FILE *stream, struct argp_state *state);

/* argp option keys */
enum {DUMMY_KEY = 129};

/* Option flags and variables.  These are initialized in parse_opt.  */

int duty;			/* --duty */
int prio;			/* --priority */
int want_verbose;		/* --verbose */

static struct argp_option options [] =
{
	{"duty", 'd', "<val>", 0, N_("Set duty to <val> milliseconds"), 0},
	{"prioity", 'p', "<prio>", 0, N_("Set SCHED_FIFO priority <prio>"), 0},
	{"verbose", 'v', NULL, 0, N_("Print more information"), 0},
	{NULL, 0, NULL, 0, NULL, 0}
};

/* The argp functions examine these global variables.  */
const char *argp_program_bug_address = "https://github.com/eurovibes/libbbmotor/issues";
void (*argp_program_version_hook) (FILE *, struct argp_state *) = show_version;

static struct argp argp =
{
	options, parse_opt, NULL,
	N_("Test Beaglebone Motor Cape stepper motors."),
	NULL, NULL, NULL
};

/* Parse a single option.  */
static error_t parse_opt (int key, char *arg __attribute__((unused)),
			  struct argp_state *state)
{
	switch (key)
	{
	case ARGP_KEY_INIT:
		/* Set up default values.  */
		duty = 500;
		prio = 90;
		want_verbose = 0;
		break;

	case 'd':			/* --duty */
		duty = atoi (arg);
		if (duty < 1 || duty > 999)
		{
			fprintf (stderr,
				 _("Duty %i out of range [1:999].\n"
				   "Reset to 500 ms.\n"),
				 duty);
			duty = 500;
		}
		break;

	case 'p':			/* --priority */
		prio = atoi (arg);
		if (prio < 1 || prio > 99)
		{
			fprintf (stderr,
				 _("Priority %i out of range [1:99].\n"
				   "Reset to 90.\n"),
				 prio);
			prio = 90;
		}
		break;

	case 'v':			/* --verbose */
		want_verbose = 1;
		break;

	case ARGP_KEY_ARG:
		argp_usage (state);
		break;

	default:
		return ARGP_ERR_UNKNOWN;
	}

	return 0;
}
void sig_han(int sig)
{
	stop = sig;
}

int main (int argc, char **argv)
{
	struct sched_param param;
	struct timespec next;
	motorcape han;
	int i, ret = 1;

	setlocale (LC_ALL, "");
	bindtextdomain (PACKAGE, LOCALEDIR);
	textdomain (PACKAGE);

	argp_parse (&argp, argc, argv, 0, NULL, NULL);

	if (want_verbose)
	{
		printf (_("%s parameter list:\n"), argv [0]);
		printf (_("Duty: %d ms\n"), duty);
		printf (_("Priority: %d\n"), prio);
	}

	mlockall (MCL_CURRENT | MCL_FUTURE);
	param.sched_priority = prio;

	if (sched_setscheduler (0, SCHED_FIFO, &param) < 0)
	{
		perror (_("Set priority failed. Abort."));
		return 1;
	}

	stop = 0;
	signal (SIGHUP, sig_han);
	signal (SIGINT, sig_han);
	signal (SIGQUIT, sig_han);

	han = motorcape_init (0x4b);
	if (!han)
	{
		perror (_("The motorcape initialization failed. Abort."));
		return 1;
	}

	if (motorcape_set_pwm (han, 10000))
	{
		perror (_("The PWM initialization failed. Abort."));
		goto out;
	}

	if (motorcape_dcm_init (han, 1))
	{
		perror (_("The DC motor initialization failed. Abort."));
		goto out;
	}

	if (clock_gettime (CLOCK_MONOTONIC, &next))
	{
		perror (_("Get current time failed. Abort"));
		goto out;
	}

	next.tv_sec += 2;
	next.tv_nsec = 0;

	i = 0;
	while (!stop)
	{
		clock_nanosleep (CLOCK_MONOTONIC, TIMER_ABSTIME, &next, NULL);

		if (motorcape_dcm_move (han, 1, 0, ++i*10))
		{
			perror (_("DC motor 1 failed to move. Abort."));
			goto out;
		}

		fprintf (stderr, N_("*"));
		next.tv_sec++;
		i %= 100;
	}

	ret = motorcape_dcm_release (han, 1);
	if (ret)
		perror (_("DC motor 1 release failed."));

out:
	motorcape_close (han);

	return ret;
}

/* Show the version number and copyright information.  */
static void show_version (FILE *stream, struct argp_state *state)
{
	(void) state;

	fputs (PACKAGE" "VERSION"\n", stream);
	fprintf (stream, _("Written by %s.\n\n"), "Benedikt Spranger");
	fprintf (stream, _("Copyright (C) %s %s\n"), "2019", "Benedikt Spranger");
	fputs(_("\
This program is free software; you may redistribute it under the terms of\n\
the GNU General Public License.  This program has absolutely no warranty.\n"),
	      stream);
}
